### PR TITLE
A default network is now created when deploying to Azure ARM

### PR DESCRIPTION
**Pull request by Graeme-Miller**
In an upcoming release of JClouds, support for Azure ARM will be added. The JClouds code for Azure ARM will place each created VM into a network based on it's group ID. As we do not use the concept of a group ID in Brooklyn, this means that each VM deployed by Brooklyn is deployed in it's own subnet. 

For most other clouds, a sensible default network is provided by the cloud. As such a Brooklyn user can be assured that if he deploys a blueprint with multiple VMs, they can reach each other on the private IPs that are provided. Unless we create a default network in Azure, those blueprints will not be portable and a user will have to manually create a network first. I believe that it is confusing for a user to have the clouds act differently on networking, which is an essential part of Brooklyn.

A few things to mention:
*) The default network/subnet is only created when needed
*) The default network/subnet is created per region
*) It is never cleaned up or deleted

--------------------------------------------------------------------
As a side note, I discovered this as AMP, which is a downstream project of Brooklyn, pulled in and was using Azure ARM support from JClouds. An older version of the JClouds code created the default network, but was recently removed meaning a regression in AMP. The code which I have added to create a default network is based on the JClouds code.

### CODE REVIEW

**Comment by drigodwin** [comment]
retest this please

**Comment by Graeme-Miller** [comment]
The failure was due to my test class having an un-approve licence. I have updated the class and pushed- this should force a retest

File: locations/jclouds/src/test/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreatorTest.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    final String TEST_SUBNET_NAME = "brooklyn-default-subnet-test-loc";
+    final String TEST_SUBNET_ID = "/test/resource/id";
+    final String TEST_LOCATION = "test-loc";
+
+    @BeforeMethod
+    public void setUp() {
+        MockitoAnnotations.initMocks(this);
+
+//
+//        when(azureComputeApi.getVirtualNetworkApi(TEST_RESOURCE_GROUP)).thenReturn(virtualNetworkApi);
```
Comment (drigodwin) [inline locations/jclouds/src/test/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreatorTest.java]: Should this be removed if it's commented?

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+            updateTemplateOptions(config, preexistingSubnet);
+            return;
+        }
+
+
+        LOG.info("Network config not specified when creating Azure location and default network/subnet does not exists. Creating");
+
+        createResourceGroupIfNeeded(api, resourceGroupName, location);
+
+        //Setup properties for creating subnet/networl
```
Comment (drigodwin) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java]: subnet/networl -> subnet/network

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    public static final Logger LOG = LoggerFactory.getLogger(DefaultAzureArmNetworkCreator.class);
+
+    private static final String DEFAULT_RESOURCE_GROUP = "brooklyn-default-resource-group";
+    private static final String DEFAULT_NETWORK_NAME = "brooklyn-default-network";
+    private static final String DEFAULT_SUBNET_NAME = "brooklyn-default-subnet";
+
+    private static final String defaultVnetAddressPrefix = "10.1.0.0/16";
+    private static final String defaultSubnetAddressPrefix = "10.1.0.0/24";
+
+    public static void createDefaultNetworkAndAddToTemplateOptions(ComputeService computeService, ConfigBag config) {
```
Comment (aledsage) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java]: I'd split this into two separate methods: `isDefaultNetworkRequired(ConfigBag)` and `createDefaultNetworkAndAddToTemplateOptions(ComputeService, ConfigBag)`.

I think it's a little confusing (given the method name) that it is a no-op if certain config options are set explicitly.

First thing in the `createDefaultNetworkAndAddToTemplateOptions`, I'd then do if (!isDefaultNetworkRequired(config)) throw new IllegalArgumentException(...)`.

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+        Map<String, Object> templateOptions = config.get(TEMPLATE_OPTIONS);
+
+        //Only create a default network if we haven't specified a network name (in template options or config) or ip options
+        if (config.containsKey(NETWORK_NAME)) {
+            LOG.info("Network config specified when creating Azure location. Not creating default network");
```
Comment (aledsage) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java]: Instead of "creating Azure location", I'd say "provisioning Azure machine".

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+public class DefaultAzureArmNetworkCreator {
+
+    public static final Logger LOG = LoggerFactory.getLogger(DefaultAzureArmNetworkCreator.class);
+
+    private static final String DEFAULT_RESOURCE_GROUP = "brooklyn-default-resource-group";
+    private static final String DEFAULT_NETWORK_NAME = "brooklyn-default-network";
+    private static final String DEFAULT_SUBNET_NAME = "brooklyn-default-subnet";
+
+    private static final String defaultVnetAddressPrefix = "10.1.0.0/16";
```
Comment (aledsage) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java]: I'd prefer all caps for constants.

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+            return;
+        }
+
+
+        LOG.info("Network config not specified when creating Azure location and default network/subnet does not exists. Creating");
+
+        createResourceGroupIfNeeded(api, resourceGroupName, location);
+
+        //Setup properties for creating subnet/network
+        Subnet subnet = Subnet.create(subnetName, null, null,
```
Comment (aledsage) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java]: I wondered whether we could move this chunk into a method like `Subnet createSubnet(api, resourceGroupName, vnetName, subnetName)`. However, having a method taking three strings in a row might well not make things clearer :-(

Reasoning is that decreasing the size of this method would make it a bit clearer.

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+        api.getVirtualNetworkApi(resourceGroupName).createOrUpdate(vnetName, location, virtualNetworkProperties);
+        Subnet createdSubnet = api.getSubnetApi(resourceGroupName, vnetName).get(subnetName);
+
+        //Add config
+        updateTemplateOptions(config, createdSubnet);
+
+    }
+
+    private static void updateTemplateOptions(ConfigBag config, Subnet createdSubnet){
+        Map<String, Object> templateOptions = config.get(TEMPLATE_OPTIONS);
```
Comment (aledsage) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java]: Don't rely on `templateOptions` from the config map being mutable. Instead, make a copy.

I'd write something like:
```
Map<String, Object> templateOptions = MutableMap.copyOf(config.get(TEMPLATE_OPTIONS))
        .put("ipOptions", ImmutableMap.of(
                "subnet", createdSubnet.id(),
                "allocateNewPublicIp", true)) // jclouds will not provide a public IP unless we set this
        .build();
config.put(TEMPLATE_OPTIONS, templateOptions);
```

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+        if(templateOptions == null) {
+            templateOptions = new HashMap<>();
+            config.put(TEMPLATE_OPTIONS, templateOptions);
+        }
+
+        Map<String, Object> ipOptions = new HashMap<>();
+        ipOptions.put("allocateNewPublicIp", true); //JClouds will not provide a public IP unless we set this
```
Comment (aledsage) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java]: Why set `allocateNewPublicIp` if the user hasn't asked for it? Is that definitely the default that we want? I probably agree, as it's more like other clouds and this is just the defaults in brooklyn (when a user isn't configuring the security options they really desire). But I worry that injecting defaults like this will be confusing - e.g. if the user supplies their own network or subnet name, then this value doesn't get set (I don't think that difference is intuitive to the user).

Comment (aledsage) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java]: Note that jclouds uses all lower-case in their branding

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java (hunk 3 above)
Comment (aledsage) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java]: I'd be tempted to add another config key that says whether to enable the automatic-default-network stuff or to skip it (where that config key defaults to true). You could declare the config key in this class, and look it up in the `ConfigBag` passed in. For example, `azureArmDefaultNetworkCreationEnabled`. Hopefully use of the word "enabled" makes it clear that setting this config key doesn't mean you'll definitely get a default network/subnet (it still depends on the other config for that vm's provisioning), but that setting it to false will always turn it off.

Comment (Graeme-Miller) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java]: re separate methods. I would prefer not to, I think it is easier to read if the logic is in one place. How about I change the name to createDefaultNetworkAndAddToTemplateOptionsIfRequired. A bit wordy, but it is clear what it is doing.

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java (hunk 6 above)
Comment (Graeme-Miller) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java]: yeah, I'm not convinced it would make things clearer. Will change though if there are any strong opinions

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java (hunk 8 above)
Comment (Graeme-Miller) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java]: setting allocateNewPublicIp makes sense to me. When a user doesn't setup any network options in brooklyn, I think they can expect:
a public IP associated with the VM
a private IP on a default network.

Setting this config gives them that. If we don't set this config, they won't be able to access their nodes (in fact, I am not sure brooklyn will be able to access the vm- i'd have to check).

Currently, without these changes, it is already confusing for a user to specify there own network as they need to know to set this option. I don't think the answer is to not set this by default, but rather to make it clearer that this needs to be set.

**Comment by drigodwin** [comment]
I'm going to merge this now, thankyou @Graeme-Miller!

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java [hunk 9]
```diff
@@ -0,0 +1,147 @@
...
+
+
+        Map<String, Object> templateOptions = config.get(TEMPLATE_OPTIONS);
+
+        //Only create a default network if we haven't specified a network name (in template options or config) or ip options
+        if (config.containsKey(NETWORK_NAME)) {
+            LOG.info("Network config specified when provisioning Azure machine. Not creating default network");
+            return;
+        }
+        if (templateOptions != null && (templateOptions.containsKey(NETWORK_NAME.getName()) || templateOptions.containsKey("ipOptions"))) {
```
Comment (neykov) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java]: The `TemplateOptions` property is named `networks`, whereas `NETWORK_NAME.getName()` is `networkName`, so it will miss user-supplied options.

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+            LOG.info("Network config specified when provisioning Azure machine. Not creating default network");
+            return;
+        }
+
+        LOG.info("Network config not specified when provisioning Azure machine. Creating default network if doesn't exist");
+
+        AzureComputeApi api = computeService.getContext().unwrapApi(AzureComputeApi.class);
+        String location = config.get(CLOUD_REGION_ID);
```
Comment (neykov) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java]: Might not be set. When using default this will be `null`.

Comment (neykov) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java]: Can only get it from  `template.getLocation()` reliably. If you move the default network creation after `template` is created you'll be able to get it. This will also make the above checks easier - all you need to do is look in the `templateOptions` object.

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java [hunk 11]
```diff
@@ -0,0 +1,147 @@
...
+
+        String resourceGroupName = DEFAULT_RESOURCE_GROUP  + "-" + location;
+        String vnetName = DEFAULT_NETWORK_NAME + "-" + location;
+        String subnetName = DEFAULT_SUBNET_NAME + "-" + location;
+
+        //Check if default already exists
+        Subnet preexistingSubnet = api.getSubnetApi(resourceGroupName, vnetName).get(subnetName);
+        if(preexistingSubnet != null){
+            LOG.info("Default Azure network and subnet already created, "+vnetName);
```
Comment (neykov) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java]: Could log that are part of the normal workflow at `debug` level.

File: locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java [hunk 12]
```diff
@@ -0,0 +1,147 @@
...
+
+        LOG.info("Network config not specified when creating Azure location and default network/subnet does not exists. Creating");
+
+        createResourceGroupIfNeeded(api, resourceGroupName, location);
+
+        //Setup properties for creating subnet/network
+        Subnet subnet = Subnet.create(subnetName, null, null,
+                Subnet.SubnetProperties.builder().addressPrefix(DEFAULT_SUBNET_ADDRESS_PREFIX).build());
+
+        VirtualNetwork.VirtualNetworkProperties virtualNetworkProperties = VirtualNetwork.VirtualNetworkProperties
```
Comment (neykov) [inline locations/jclouds/src/main/java/org/apache/brooklyn/location/jclouds/networking/creator/DefaultAzureArmNetworkCreator.java]: The vnet might exist, but subnet could be deleted for whatever reason. I'd split this in two steps, doing the exists check for each one.

**Comment by neykov** [comment]
The most important from above comments is that `location` (which is region in brooklyn speak) could not be set, using the jclouds' default. Would be useful to have that improved.
Another thing that I think is important is locking. When deploying a large application it's almost certain that the logic will get confused because everyone is trying to create the default network in parallel.
